Document minimal operational capabilities. (#93)
abetterinternet/prio-documents: draft-pda-protocol.md

The leader responds to well-formed requests to `[leader]/upload` with status 200
and an empty body. Malformed requests are handled as described in
{{pa-error-common-aborts}}. Clients MUST NOT upload the same measurement value in

> more than one report if the leader responds with status 200 and an empty body.

Reviewer: It seems like this would require clients to maintain a list of uploaded reports to avoid double-uploads, which contradicts the later statement:
>Clients are not expected to store any state across any upload flows

Besides which, I think leaders will need to implement detection of duplicate report uploads no matter what, in which case they can return an error to clients in the case of a double upload, which downgrades this to a SHOULD NOT.

Author: Yeah, this seems reasonable.

[TODO: Define Hits-specific protocol messages.]

# Operational Considerations {#operational-capabilities}

> Prio has inherent constraints derived from the tradeoff between privacy

Reviewer 2: ```suggestion
PDA protocols have inherent constraints derived from the tradeoff between privacy
```

guarantees and computational complexity. These tradeoffs influence how
applications may choose to utilize services implementing the specification.

## Protocol participant capabilities {#entity-capabilities}

The design in this document has different assumptions and requirements for
different protocol participants, including clients, aggregators, and
collectors. This section describes these capabilities in more detail.

### Client capabilities

Clients have limited capabilities and requirements. Their only inputs to the protocol

> are (1) the PDAParam structure configured out of band and (2) a corresponding value to be

Reviewer: Corresponding to what?

> are (1) the PDAParam structure configured out of band and (2) a corresponding value to be
> submitted in a report. Clients are not expected to store any state across any upload

Reviewer 2: ```suggestion
are (1) the PDAParam structure configured out of band and (2) a measurement. Clients
are not expected to store any state across any upload
```

> flows, nor are required to implement any sort of report retry mechanism should it fail.

Reviewer: ```suggestion
flows, nor are they required to implement any sort of report upload retry mechanism.
```
It was unclear what "it" referred to, and I think retry implies a failure.

By design, the protocol in this document is robust against individual client upload

> failures since the protocol output is an aggregate over all inputs.

Reviewer 2: Also worth mentioning here: The aggregators validate inputs before consuming them.

Author: That's just par for the course -- I don't think it changes any of the capacity or operational requirements?

### Aggregator capabilities

Helpers and leaders have different operational requirements. The design in this
document assumes an operationally competent leader, i.e., one that has no storage

> or computation limitations or constraints, and minimalistic helper, i.e., one that

Reviewer 2: ```suggestion
or computation limitations or constraints, but only a modestly provisioned helper, i.e., one that
```

has computation, bandwidth, and storage constraints. By design, leaders must be
at least as capable as helpers, where helpers are generally required to:

- Support the collect protocol, which includes verifying and aggregating
  sets of reports in a given batch; and
- Publish and manage an HPKE configuration that can be used for the upload protocol;

In addition, for each PDAParam instance, helpers are required to:

- Implement some form of batch-to-report and intra-batch replay mitigation storage,
  which includes some way of tracking batch report size, with optional support for

>   state offloading.

Reviewer 2: Clarify here that some of the offloaded state can be used for anti-replay, but not all.

Author: I don't think that's necessary to do here.

> [OPEN ISSUE: describe intra- and inter-batch replay attacks in the security considerations]

Reviewer 2: This will be resolved (at least partially) by #90.

Author: Yep, noted it as such.

> Beyond the minimal capabiltiies required of helpers, leaders are generally required to:

Reviewer: ```suggestion
Beyond the minimal capabilities required of helpers, leaders are generally required to:
```

- Support the upload protocol and store client reports for a given PDAParam instance,
  where each report maps uniquely to a single batch, and index this storage by batch durations;
- Track batch report size during each collect flow and request encrypted output shares
  from helpers.

In addition, for each PDAParam instance, leaders are required to:

- Ensure that, for each report, all aggregator shares are either included or not in a

>   given batch;

Reviewer: I find this a bit confusing. Is the intent here that for a report, the corresponding aggregator *output* shares must either both be included or both be omitted?

Reviewer 2: +1

Author: This is indeed somewhat confusing, and probably not needed, so dropped.

> - Implement and store state for some form of inter- and intra-batch replay mitigation; and

Reviewer 2: This is also a requirement for the helper.

- Store helper state associated with a given PDAParam batch.

### Collector capabilities

Collectors statefully interact with aggregators to produce an aggregate output. Their
input to the protocol is the PAParam structure, configured out of band, which contains
the corresponding batch window and size. For each collect invocation, collectors are
required to keep state from the start of the protocol to the end as needed to produce

> the final aggregate output.

Reviewer 2: They also need to keep around an HPKE secret key for the lifetime of the PDAParam.

different protocol participants, including clients, aggregators, and
collectors. This section describes these capabilities in more detail.

### Client capabilities

Clients have limited capabilities and requirements. Their only inputs to the protocol
are (1) the PDAParam structure configured out of band and (2) a measurement. Clients
are not expected to store any state across any upload
flows, nor are they required to implement any sort of report upload retry mechanism.
By design, the protocol in this document is robust against individual client upload
failures since the protocol output is an aggregate over all inputs.

### Aggregator capabilities

Helpers and leaders have different operational requirements. The design in this
document assumes an operationally competent leader, i.e., one that has no storage
or computation limitations or constraints, but only a modestly provisioned helper, i.e., one that
has computation, bandwidth, and storage constraints. By design, leaders must be
at least as capable as helpers, where helpers are generally required to:

- Support the collect protocol, which includes verifying and aggregating
  sets of reports in a given batch; and

> - Publish and manage an HPKE configuration that can be used for the upload protocol;

Reviewer 2: ```suggestion
- Publish and manage an HPKE configuration that can be used for the upload protocol.
```

In addition, for each PDAParam instance, helpers are required to:

- Implement some form of batch-to-report index, as well as inter- and intra-batch
  replay mitigation storage, which includes some way of tracking batch report size
  with optional support for state offloading. Some of this state may be used for

>   replay attack mitigation. One relay mitigation strategy is described in {{anti-replay}}.

Reviewer 2: ```suggestion
  replay attack mitigation. One replay mitigation strategy is described in {{anti-replay}}.
```
Also, isn't the strategy in {{anti-replay}} *the* strategy? Do we intend to allow deployments to skip this, or do something else?

Beyond the minimal capabilities required of helpers, leaders are generally required to:

- Support the upload protocol and store client reports for a given PDAParam instance,
  where each report maps uniquely to a single batch, and index this storage by batch durations;
- Track batch report size during each collect flow and request encrypted output shares
  from helpers.

In addition, for each PDAParam instance, leaders are required to:

- Ensure that, for each report, all aggregator shares are either included or not in a
  given batch;

> - Implement and store state for some form of inter- and intra-batch replay mitigation; and

Reviewer 2: Again, at this point it seems like there is one and only one replay mitigation, and it's spelled out in the doc.

Author: (Commented below.)

- Store helper state associated with a given PDAParam batch.

### Collector capabilities

Collectors statefully interact with aggregators to produce an aggregate output. Their

> input to the protocol is the PAParam structure, configured out of band, which contains

Reviewer: ```suggestion
input to the protocol is the PDAParam structure, configured out of band, which contains
```

1. Minimum batch size. The aggregation batch size has an obvious impact on
   privacy. (A batch size of one hides nothing of the input.)
2. Aggregation function choice. Some aggregation functions leak slightly more
   than the function output itself.

[TODO: discuss these in more detail.]

## Infrastructure diversity

> Prio deployments should ensure that aggregators do not have common dependencies

Reviewer 2: ```suggestion
PDA deployments should ensure that aggregators do not have common dependencies
```

that would enable a single vendor to reassemble inputs. For example, if all
participating aggregators stored unencrypted input shares on the same cloud
object storage service, then that cloud vendor would be able to reassemble all
the input shares and defeat privacy.

## Data integrity constraints

Data integrity concerns the accuracy and correctness of the outputs in the
system. The integrity of the output can be influenced by an incomplete round of
aggregation caused by network partitions, or by bad actors attempting to cause
inaccuracies in the aggregates. An example data integrity constraint is that

> every share must be processed exactly once by all aggregators. Data integrity

Reviewer: Distributed systems people will hiss at you for saying "exactly once". Should this be "at most once"?

Author: Hah, yeah, fair :-) That's a good suggestion!

constraints may be at odds with the threat model if meeting the constraints

> requires replaying data.

Reviewer 2: > Data integrity constraints may be at odds with the threat model if meeting the constraints requires replaying data.

I realize this statement was copy-pated, but I think we should remove it. The origin of this sentence is the observation that Hits requires multiple collect rounds over the same data. This fact isn't at odds with the need for anti-replay, since mutliple collect requests aren't "replaying" reports in an illegal way.

Author: Less text is better :-) I'll remove it.

Aggregator operators should expect to encounter invalid inputs during regular
operation due to misconfigured or malicious clients. Low volumes of errors are

> tolerable; the input-verification protocol and AFEs are robust in the face of

Reviewer 2: ```suggestion
tolerable; the input-verification procedure of the PDA protocol
```
The term "AFE" is specific to Prio.

Author: This is moved text, so let's change that separately.

> tolerable; the input-verification protocol and AFEs are robust in the face of

Reviewer: "AFE" turns up in a handful of places in the doc now. Perhaps it merits an entry in the `Terminology` section at the top, especially since it's not obvious that "AFE" expands to "affine-aggregatable encodings".

Reviewer 2: In fact, I'm not so sure we want to bolt ourselves to this term too much. First, it's not meaningful for every PDA protocol (for Hits in particular). Second, the term was coined in the original Prio paper, which we don't actually implement.

Author: (Overcome by events)

malformed data. Aggregators may need to detect and mitigate statistically
significant floods of invalid or identical inputs that affect accuracy, e.g.,
denial of service (DoS) events.

Certain classes of errors do not exist in the input-validation protocol
considered in this document. For example, packet loss errors when clients make
requests directly to aggregators are not relevant when the leader proxies

> requests and controls the schedule for signaling aggregation rounds.

Reviewer 2: I realize this is copy-pasted, but I'm questioning the value of this paragraph at this point. Is it saying anything that's new and not obvious?

Author: Nope, we can delete.
